cpp/src/arrow/dataset/file_skyhook.cc

```diff
@@ -0,0 +1,280 @@
+  if (connected) {
+    return Status::OK();
+  }
+
+  // Locks the mutex. Only one thread can pass here at a time.
+  // Another thread handled the connection already.
+  std::unique_lock<std::mutex> lock(connection_mutex);
+  if (connected) {
+    return Status::OK();
+  }
+  connected = true;
+
+  if (rados->init2(ctx.user_name.c_str(), ctx.cluster_name.c_str(), 0))
+    return Status::Invalid("librados::init2 returned non-zero exit code.");
+
+  if (rados->conf_read_file(ctx.ceph_config_path.c_str()))
+    return Status::Invalid("librados::conf_read_file returned non-zero exit code.");
+
+  if (rados->connect())
+    return Status::Invalid("librados::connect returned non-zero exit code.");
+
+  if (rados->ioctx_create(ctx.data_pool.c_str(), ioCtx))
+    return Status::Invalid("librados::ioctx_create returned non-zero exit code.");
+
+  return Status::OK();
+}
+
+Status RadosConnection::Shutdown() {
+  rados->shutdown();
+  return Status::OK();
+}
+
+}  // namespace connection
+
+/// \brief A ScanTask to scan a file fragment in Skyhook format.
+class SkyhookScanTask : public ScanTask {
+ public:
+  SkyhookScanTask(std::shared_ptr<ScanOptions> options,
+                  std::shared_ptr<Fragment> fragment, FileSource source,
+                  std::shared_ptr<SkyhookDirectObjectAccess> doa, int fragment_format)
+      : ScanTask(std::move(options), std::move(fragment)),
+        source_(std::move(source)),
+        doa_(std::move(doa)),
+        fragment_format_(fragment_format) {}
+
+  Result<RecordBatchIterator> Execute() override {
+    struct stat st {};
+    ARROW_RETURN_NOT_OK(doa_->Stat(source_.path(), st));
+
+    ceph::bufferlist request;
+    ARROW_RETURN_NOT_OK(
+        SerializeScanRequest(options_, fragment_format_, st.st_size, request));
+
+    ceph::bufferlist result;
+    ARROW_RETURN_NOT_OK(doa_->Exec(st.st_ino, "scan_op", request, result));
+
+    RecordBatchVector batches;
+    ARROW_RETURN_NOT_OK(DeserializeTable(batches, result, !options_->use_threads));
+    return MakeVectorIterator(batches);
+  }
+
+ protected:
+  FileSource source_;
+  std::shared_ptr<SkyhookDirectObjectAccess> doa_;
+  int fragment_format_;
+};
+
+SkyhookFileFormat::SkyhookFileFormat(const std::string& fragment_format,
+                                     const std::string& ceph_config_path,
+                                     const std::string& data_pool,
+                                     const std::string& user_name,
+                                     const std::string& cluster_name,
+                                     const std::string& cls_name)
+    : SkyhookFileFormat(std::make_shared<connection::RadosConnection>(
+          connection::RadosConnection::RadosConnectionCtx(
+              ceph_config_path, data_pool, user_name, cluster_name, cls_name))) {
+  fragment_format_ = fragment_format;
+}
+
+SkyhookFileFormat::SkyhookFileFormat(
+    const std::shared_ptr<connection::RadosConnection>& connection) {
+  connection->Connect();
```

Comment:
I'm not sure what resources are consumed when calling `Connect` but generally I wouldn't expect any connections to be opened just be creating a `SkyhookFileFormat` instance.  It might make more sense to establish the connection on `Inspect` or `ScanFile`.